src/EpochIOGroup.cpp

```diff
@@ -0,0 +1,397 @@
+        for (const auto &kv : m_valid_signal_name) {
+            names.insert(kv.first);
+        }
+        return names;
+    }
+
+    std::set<std::string> EpochIOGroup::control_names(void) const
+    {
+        return {};
+    }
+
+    bool EpochIOGroup::is_valid_signal(const std::string &signal_name) const
+    {
+        return m_valid_signal_name.find(signal_name) != m_valid_signal_name.end();
+    }
+
+    bool EpochIOGroup::is_valid_control(const std::string &control_name) const
+    {
+        return false;
+    }
+
+    int EpochIOGroup::signal_domain_type(const std::string &signal_name) const
+    {
+        int result = GEOPM_DOMAIN_INVALID;
+        if (is_valid_signal(signal_name)) {
+            result = GEOPM_DOMAIN_CPU;
+        }
+        return result;
+    }
+
+    int EpochIOGroup::control_domain_type(const std::string &control_name) const
+    {
+        return GEOPM_DOMAIN_INVALID;
+    }
+
+    int EpochIOGroup::push_signal(const std::string &signal_name, int domain_type, int domain_idx)
+    {
+        if (!is_valid_signal(signal_name)) {
+            throw Exception("EpochIOGroup::push_signal(): signal_name " + signal_name +
+                            " not valid for EpochIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        check_domain(domain_type, domain_idx);
+        if (m_is_batch_read) {
+            throw Exception("EpochIOGroup::push_signal(): cannot push signal after call to read_batch().",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        int signal_type = signal_name_to_type(signal_name);
+        // check if already pushed
+        int result = -1;
+        auto it = m_signal_idx_map.find({signal_type, domain_idx});
+        if (it != m_signal_idx_map.end()) {
+            result = it->second;
+        }
+        if (result == -1) {
+            result = m_active_signal.size();
+            m_active_signal.push_back({signal_type, domain_type, domain_idx});
+            // m_is_signal_pushed = true;
+            m_signal_idx_map[{signal_type, domain_idx}] = result;
+        }
+        return result;
+    }
+
+    int EpochIOGroup::push_control(const std::string &control_name, int domain_type, int domain_idx)
+    {
+        throw Exception("EpochIOGroup::push_control(): there are no controls supported by the EpochIOGroup",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+    }
+
+    void EpochIOGroup::read_batch(void)
+    {
+        if (!m_is_initialized) {
+            init();
+        }
+        /// @todo: assume update_records() will get called by controller
+        auto records = m_app.get_records();
+        for (const auto &record : records) {
+            if (m_epoch_map.find(record.process) == m_epoch_map.end()) {
+                /// @todo: this should not happen; for now skip record
+                continue;
+            }
+            m_epoch_map.at(record.process)->update(record);
```

Comment:
keep a local copy of the m_epoch_map iterator and check if process hasn't changed before doing map look up. Of questionable value.  Use godbolt and don't be a smart ass.